### PR TITLE
Experimental extraction algorithms

### PR DESCRIPTION
**Pull request by TalosOfCrete**
Within this PR is an implementation of [RetinaFace](https://github.com/biubug6/Pytorch_Retinaface/) and [Adaptive Wing Loss FAN](https://github.com/protossw512/AdaptiveWingLoss).  Both of these algorithms have dramatic gains in accuracy over S3FD and standard FAN respectively, and would be of great use to the users of this repository.  The code as it stands right now does not produce any errors when attempting to execute, and the PyTorch code does not conflict with any of the TensorFlow code; however, there are some problems that I could use some help rectifying:

1. It appears that the system for face bounding box detection that I have set up (which follows the GitHub link above) currently fails to produce any face detections.  This is strange, as an earlier version of this code (which I spent a few hours today trying to remember) resulted in the algorithm not only correctly identifying the number of bounding boxes per frame, but significantly outperforming S3FD: when compared, S3FD would produce 5-6 false positives with low (~5%) confidence, while RetinaFace would produce 2 correct positives with much higher (~35%) confidence.  Upon examining the debug frames, it also appeared that it not only found bounding boxes for frames that S3FD completely missed, it found quite good bounding boxes overall.  Here are some examples (I can provide the original frames to prove that S3FD would underperform/fail on them if you so desire):

The above frame (and the ones very similar to it) were not picked up at all by S3FD.  Note the dark setting coupled with an obfuscation, and the accuracy of the blue bounding box.

Yet another frame (and the ones similar to it) that were too dark and blurry for S3FD, but RetinaFace had no problem with.

**An example of the bug referenced above:**

While I attempted to recreate the code that produced the example frames, I kept running into dead ends.  I am hoping that putting up this PR and #760 will encourage further experimentation to make it functional again; and, to that end, functioning optimally within DFL.

2. It is currently untested on GPUs, as my colleagues at my university's HPC center are currently under the gun of a NeurIPS deadline.  Hence, I was only ever able to test it on CPUs (as their need for GPUs at this point in time far outweighs mine).  Therefore, it is imperative that this is tested for device compatibility as well.

While this PR sits, I will continue to mess with the code to try to get it in working order again (to be able to reproduce the above examples).  However, attempting to find the bug has turned into quite the psychological grind, so my progress is slowing.  Further experimentation by people who are more familiar with DFL's bounding box and landmark systems will likely find the issue(s) with the code much more quickly than I.

### CODE REVIEW

**Comment by iperov** [comment]
where is RetinaFaceExtractor.py ?

**Comment by TalosOfCrete** [comment]
It’s all in Facelib. Again, it currently doesn’t properly find the bounding boxes within DFL’s framework, so I was hoping someone with much more knowledge about DFL than I could help debug it.

**Comment by iperov** [comment]
plz stop wasting my time 
firefox_2020-06-02_11-35-32

**Comment by iperov** [comment]
There are a lot of files for retinaface. I don't like such redundant solutions.
This is how pytorch researchers are described : nothing common with open source programming and elegant code, which can be understood by users.
Look at S3FD or FAN, I ported it to minimal single py file without tonns of shit as in original repos.

also no pth file. It has size > 100mb, so you cannot upload it to github.

**Comment by TalosOfCrete** [comment]
Of course @iperov .  Yes, the bug is in RetinaFace.  I have also added the related files, and am providing .pth files [here](https://drive.google.com/drive/folders/1OE2htKGsLffPuzbUJwVV2BVQeYVw1s2A?usp=sharing)

Based on what you are currently saying, what would you say is the wisest course of action?

**Comment by iperov** [comment]
wisest course of action : do nothing.

**Comment by TalosOfCrete** [comment]
@iperov You sure?  I believe it’s worth the effort to gain dramatic improvements in extraction robustness.

**Comment by iperov** [comment]
it does not.
Missed frames can be extracted manually.
also XSeg is manual job. If you want quality deepfake you have to do the manual job anyway.

**Comment by TalosOfCrete** [comment]
But we’re still using the landmark detector in manual regardless, right?  Having a higher quality landmark detector would certainly be very useful, and a more robust face detector would reduce the amount of manual fixing required.  A win for everyone.

**Comment by iperov** [comment]
no
https://github.com/iperov/DeepFaceLab/issues/760#issuecomment-633822497

**Comment by TalosOfCrete** [comment]
That’s understood.  My main point is that we can get better face alignment data for the user to then mask, and it really wouldn’t take that much effort.
